controlplane/kubeadm/internal/controllers/status.go

```diff
@@ -135,3 +142,257 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, contro
+		messages = append(messages, "waiting for etcd cluster to be healthy")
+	}
+
+	v1beta2conditions.Set(kcp, metav1.Condition{
+		Type:    controlplanev1.KubeadmControlPlaneScalingUpV1Beta2Condition,
+		Status:  metav1.ConditionTrue,
+		Reason:  controlplanev1.KubeadmControlPlaneScalingUpV1Beta2Reason,
+		Message: strings.Join(messages, "; "),
+	})
+}
+
+func setScalingDownCondition(_ context.Context, kcp *controlplanev1.KubeadmControlPlane, machines collections.Machines, preflightChecks internal.PreflightCheckResults) {
+	currentReplicas := kcp.Status.Replicas
+	desiredReplicas := ptr.Deref(kcp.Spec.Replicas, 0)
+	if !kcp.DeletionTimestamp.IsZero() {
+		desiredReplicas = 0
+	}
+
+	if currentReplicas <= desiredReplicas {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:   controlplanev1.KubeadmControlPlaneScalingDownV1Beta2Condition,
+			Status: metav1.ConditionFalse,
+			Reason: controlplanev1.KubeadmControlPlaneNotScalingDownV1Beta2Reason,
+		})
+		return
+	}
+
+	messages := []string{fmt.Sprintf("ScalingDown from %d to %d replicas", currentReplicas, desiredReplicas)}
+	if preflightChecks.HasDeletingMachine {
+		messages = append(messages, "Waiting for Machine being deleted")
+	}
+
+	machinesStaleDeleting := []string{}
+	for _, machine := range machines {
+		if !machine.GetDeletionTimestamp().IsZero() && time.Since(machine.GetDeletionTimestamp().Time) >= time.Minute*30 {
+			machinesStaleDeleting = append(machinesStaleDeleting, machine.GetName())
+		}
+	}
+
+	if len(machinesStaleDeleting) == 1 {
+		messages = append(messages, fmt.Sprintf("Machine %s is in deletion by more than 30 minutes", strings.Join(machinesStaleDeleting, ", ")))
+	} else if len(machinesStaleDeleting) > 1 {
+		messages = append(messages, fmt.Sprintf("Machines %s are in deletion by more than 30 minutes", strings.Join(machinesStaleDeleting, ", ")))
+	}
+
+	if preflightChecks.ControlPlaneComponentsNotHealthy {
+		messages = append(messages, "Waiting for control plane components to be healthy")
+	}
+
+	if preflightChecks.EtcdClusterNotHealthy {
+		messages = append(messages, "Waiting for etcd cluster to be healthy")
+	}
+
+	v1beta2conditions.Set(kcp, metav1.Condition{
+		Type:    controlplanev1.KubeadmControlPlaneScalingDownV1Beta2Condition,
+		Status:  metav1.ConditionTrue,
+		Reason:  controlplanev1.KubeadmControlPlaneScalingDownV1Beta2Reason,
+		Message: strings.Join(messages, "; "),
+	})
+}
+
+func setMachinesReadyCondition(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines collections.Machines) {
+	if len(machines) == 0 {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:   controlplanev1.KubeadmControlPlaneMachinesReadyV1Beta2Condition,
+			Status: metav1.ConditionTrue,
+			Reason: controlplanev1.KubeadmControlPlaneMachinesReadyNoReplicasV1Beta2Reason,
+		})
+		return
+	}
+
+	readyCondition, err := v1beta2conditions.NewAggregateCondition(
+		machines.UnsortedList(), clusterv1.MachineReadyV1Beta2Condition,
```

Comment:
Is NewAggregateCondition sorting the machines internally? (took a quick look, but I'm not sure)

(same for up-to-date & remediating)

Reply:
NewAggregateCondition: has different layers of sort (by kind, by message - the one reported by more object get first-, and now also by object name 😉